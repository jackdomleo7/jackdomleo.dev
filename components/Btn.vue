<template>
  <button v-if="!href" class="btn" :class="{ 'btn--square': square }" :type="type" @click="$emit('click')">
    <slot />
  </button>
  <a v-else-if="href.startsWith('https://') || href.startsWith('#')" class="btn" :class="{ 'btn--square': square }" :href="href" target="_blank" @click="$emit('click')">
    <slot />
  </a>
  <nuxt-link v-else class="btn" :class="{ 'btn--square': square }" :to="href" @click="$emit('click')">
    <slot />
  </nuxt-link>
</template>

<script lang="ts">
import Vue from 'vue'

export default Vue.extend({
  name: 'Btn',
  props: {
    href: {
      type: String,
      default: undefined
    },
    square: {
      type: Boolean,
      default: false
    },
    type: {
      type: String,
      default: 'button',
      validator: (value: string): boolean => {
        return ['button', 'submit', 'reset'].includes(value)
      }
    }
  }
})
</script>

<style lang="scss" scoped>
.btn {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  background-color: var(--colour-primary);
  color: var(--colour-text-button);
  border-radius: 1.5rem;
  border: 1px solid var(--colour-primary);
  box-shadow: var(--shadow-standard);
  text-decoration: none;
  font-size: var(--text-large);
  font-weight: 700;
  padding: 0.75rem 2.5rem;
  height: 3rem;
  max-width: 100%;
  cursor: pointer;
  transition: var(--standard-animation-timing);

  @media (min-width: $responsive-standard-tablet) {
    min-width: 15.625rem;
  }

  &--square {
    border-radius: 0.1875rem;
    padding-inline: 1.5rem;
  }

  &:active {
    box-shadow: none;
  }

  &:disabled {
    opacity: 0.7;
    cursor: default;
  }
}
</style>
