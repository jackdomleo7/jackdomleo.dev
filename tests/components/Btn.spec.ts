import { mount, shallowMount } from '@vue/test-utils'
import { configureAxe, toHaveNoViolations } from 'jest-axe'
import Btn from '@/components/Btn.vue'

expect.extend(toHaveNoViolations)

const axe = configureAxe({
  rules: {
    region: { enabled: false } // Don't need to check because components are loaded in isolation
  }
})

describe('Btn', () => {
  /**
   * Vue instance
   */

  it('is a Vue instance', () => {
    const wrapper = mount(Btn, {
      slots: {
        default: 'Click me'
      }
    })

    expect(wrapper.vm).toBeTruthy()
  })

  /**
   * Snapshots
   */

  it('matches snapshot when mounted', () => {
    const wrapper = mount(Btn, {
      slots: {
        default: 'Click me'
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('matches snapshot when shallow mounted', () => {
    const wrapper = shallowMount(Btn, {
      slots: {
        default: 'Click me'
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  /**
   * A11y
   */

  it('to have no obvious accessibility violations', async () => {
    const wrapper = mount(Btn, {
      slots: {
        default: 'Click me'
      }
    })

    await expect(axe(wrapper.element)).resolves.toHaveNoViolations()
  })

  /**
   * Emitters
   */

  it('emits the "click" when the <button> is clicked', async () => {
    const wrapper = shallowMount(Btn, {
      slots: {
        default: 'Click me'
      }
    })

    expect(wrapper.find('button.btn').exists()).toBe(true)

    await wrapper.find('button.btn').trigger('click')

    expect(wrapper.emitted('click')).toBeTruthy()
  })

  it('emits the "click" when the <a> is clicked', async () => {
    const wrapper = shallowMount(Btn, {
      slots: {
        default: 'Click me'
      },
      propsData: {
        href: '#'
      }
    })

    expect(wrapper.find('a.btn').exists()).toBe(true)

    await wrapper.find('a.btn').trigger('click')

    expect(wrapper.emitted('click')).toBeTruthy()
  })

  it('emits the "click" when the <nuxt-link> is clicked', async () => {
    const wrapper = mount(Btn, {
      slots: {
        default: 'Click me'
      },
      propsData: {
        href: '/blog'
      },
      stubs: ['nuxt-link']
    })

    expect(wrapper.find('nuxt-link-stub.btn').exists()).toBe(true)

    await wrapper.find('.btn').trigger('click')

    // TODO: For some reason, nuxt-link doesn't emit 'click' is tests but does in the live site
    // expect(wrapper.emitted('click')).toBeTruthy()
  })

  /**
   * Props
   */

  it('renders a <button> when no href is set', () => {
    const wrapper = shallowMount(Btn, {
      slots: {
        default: 'Click me'
      }
    })

    expect(wrapper.find('button.btn').exists()).toBe(true)
  })

  it('renders an <a> when href is set to an external link', () => {
    const wrapper = shallowMount(Btn, {
      slots: {
        default: 'Click me'
      },
      propsData: {
        href: 'https://jackdomleo.dev'
      }
    })

    expect(wrapper.find('a.btn').exists()).toBe(true)
  })

  it('renders an <a> when href is set to a link on the same page', () => {
    const wrapper = shallowMount(Btn, {
      slots: {
        default: 'Click me'
      },
      propsData: {
        href: '#about'
      }
    })

    expect(wrapper.find('a.btn').exists()).toBe(true)
  })

  it('renders a <nuxt-link> when href is set to an internal link', () => {
    const wrapper = shallowMount(Btn, {
      slots: {
        default: 'Click me'
      },
      propsData: {
        href: '/blog'
      },
      stubs: ['nuxt-link']
    })

    expect(wrapper.find('nuxt-link-stub.btn').exists()).toBe(true)
  })

  it('renders the "btn--square" class when square is set for a <button>', () => {
    const wrapper = shallowMount(Btn, {
      slots: {
        default: 'Click me'
      },
      propsData: {
        square: true
      }
    })

    expect(wrapper.find('button.btn').classes()).toContain('btn--square')
  })

  it('renders the "btn--square" class when square is set for a <a>', () => {
    const wrapper = shallowMount(Btn, {
      slots: {
        default: 'Click me'
      },
      propsData: {
        href: 'https://jackdomleo.dev',
        square: true
      }
    })

    expect(wrapper.find('a.btn').classes()).toContain('btn--square')
  })

  it('renders the "btn--square" class when square is set for a <nuxt-link>', () => {
    const wrapper = shallowMount(Btn, {
      slots: {
        default: 'Click me'
      },
      propsData: {
        href: '/blog',
        square: true
      },
      stubs: ['nuxt-link']
    })

    expect(wrapper.find('nuxt-link-stub.btn').classes()).toContain('btn--square')
  })

  it('renders a <button type="button"> when the type is unset', () => {
    const wrapper = shallowMount(Btn, {
      slots: {
        default: 'Click me'
      }
    })

    expect(wrapper.find('button.btn').attributes('type')).toBe('button')
  })

  it('renders a <button type="button"> when the type is set to "button"', () => {
    const wrapper = shallowMount(Btn, {
      slots: {
        default: 'Click me'
      },
      propsData: {
        type: 'button'
      }
    })

    expect(wrapper.find('button.btn').attributes('type')).toBe('button')
  })

  it('renders a <button type="submit"> when the type is set to "submit"', () => {
    const wrapper = shallowMount(Btn, {
      slots: {
        default: 'Click me'
      },
      propsData: {
        type: 'submit'
      }
    })

    expect(wrapper.find('button.btn').attributes('type')).toBe('submit')
  })

  it('renders a <button type="reset"> when the type is set to "reset"', () => {
    const wrapper = shallowMount(Btn, {
      slots: {
        default: 'Click me'
      },
      propsData: {
        type: 'reset'
      }
    })

    expect(wrapper.find('button.btn').attributes('type')).toBe('reset')
  })
})
