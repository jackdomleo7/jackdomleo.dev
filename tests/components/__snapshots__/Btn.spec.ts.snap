// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Btn matches snapshot when mounted 1`] = `
<button
  class="btn"
  type="button"
>
  Click me
</button>
`;

exports[`Btn matches snapshot when shallow mounted 1`] = `
<button
  class="btn"
  type="button"
>
  Click me
</button>
`;
